x/subscription/keeper/subscription.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/lavanet/lava/x/subscription/types"
+)
+
+// SetSubscription sets a subscription (of a consumer) in the store
+func (k Keeper) SetSubscription(ctx sdk.Context, sub types.Subscription) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.SubscriptionKeyPrefix))
+	b := k.cdc.MustMarshal(&sub)
+	store.Set(types.SubscriptionKey(sub.Consumer), b)
+}
+
+// GetSubscription returns the subscription of a given consumer
+func (k Keeper) GetSubscription(ctx sdk.Context, consumer string) (val types.Subscription, found bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.SubscriptionKeyPrefix))
+
+	b := store.Get(types.SubscriptionKey(consumer))
+	if b == nil {
+		return val, false
+	}
+
+	k.cdc.MustUnmarshal(b, &val)
+	return val, true
+}
+
+// RemoveSubscription removes the subscription (of a consumer) from the store
+func (k Keeper) RemoveSubscription(ctx sdk.Context, consumer string) {
+	sub, _ := k.GetSubscription(ctx, consumer)
+	packages.PutPackage(ctx, sub.PackageIndex, sub.Block)
+
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.SubscriptionKeyPrefix))
+	store.Delete(types.SubscriptionKey(consumer))
+}
+
+// GetAllSubscription returns all subscriptions that satisfy the condition
+func (k Keeper) GetCondSubscription(ctx sdk.Context, cond func(sub types.Subscription) bool) (list []types.Subscription) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.SubscriptionKeyPrefix))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.Subscription
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		if cond == nil || cond(val) {
+			list = append(list, val)
+		}
+	}
+
+	return
+}
+
+// GetAllSubscription returns all subscription (of all consumers)
+func (k Keeper) GetAllSubscription(ctx sdk.Context) []types.Subscription {
+	return k.GetCondSubscription(ctx, nil)
+}
+
+func endOfMonth(date time.Time) time.Time {
+	return date.AddDate(0, 1, -date.Day())
+}
+
+func leapYear(date time.Time) bool {
+	year := date.Year()
+	return year%400 == 0 || (year%4 == 0 && year%100 != 0)
+}
+
+func nextMonth(date time.Time) time.Time {
+	next := date.AddDate(0, 1, 0)
+
+	if date.Day() == endOfMonth(date).Day() {
+		next := time.Date(
+			date.Year(),
+			date.Month()+1,
+			1,
+			date.Hour(),
+			date.Minute(),
+			date.Second(),
+			date.Nanosecond(),
+			time.UTC)
+		return endOfMonth(next)
+	}
+
+	return next
+}
+
+func nextYear(date time.Time) time.Time {
+	next := date.AddDate(1, 0, 0)
+
+	if date.Month() == 2 {
+		if date.Day() == 29 {
+			next = date.AddDate(1, 0, -1)
+		}
+		if date.Day() == 28 && leapYear(date) {
+			next = date.AddDate(1, 0, 1)
+		}
+	}
+
+	return next
+}
+
+// CreateSubscription creates a subscription for a consumer
+func (k Keeper) CreateSubscription(ctx sdk.Context, sub types.Subscription) error {
```

Comment:
who creates the "sub" argument? since this function get a subscription object, verifies it and sets it to the store, maybe call the function "VerifyAndSetSubscription"? or "SetSubscription" (the original "SetSubscription" will turn to "setSubscription").
The current name is kind of confusing since this function doesn't create a subscription object.

Reply:
It gets a partially filled Subscription object - with 4 fields: creator, consumer, plan-index, is-yearly; And it fills the rest. Will change it to accept the arguments explicitly.